scalapact-core/src/main/scala/com/itv/scalapactcore/common/matching/InteractionMatchingPrograms.scala

```diff
@@ -0,0 +1,213 @@
+package com.itv.scalapactcore.common.matching
+
+import com.itv.scalapactcore.common.matching.PathMatching.PathAndQuery
+
+import scalaz._
+import Scalaz._
+import com.itv.scalapactcore.{InteractionRequest, InteractionResponse}
+
+sealed trait MatchStatusOutcome
+case object MatchStatusSuccess extends MatchStatusOutcome
+case object MatchStatusFailed extends MatchStatusOutcome
+
+sealed trait MatchHeadersOutcome
+case object MatchHeadersSuccess extends MatchHeadersOutcome
+case object MatchHeadersFailed extends MatchHeadersOutcome //TODO: What kind of match failures do we get that we can model?
+
+sealed trait MatchBodyOutcome
+case object MatchBodySuccess extends MatchBodyOutcome
+case object MatchBodyFailed extends MatchBodyOutcome //TODO: What kind of match failures do we get that we can model?
+
+sealed trait MatchMethodOutcome
+case object MatchMethodSuccess extends MatchMethodOutcome
+case object MatchMethodFailed extends MatchMethodOutcome //TODO: What kind of match failures do we get that we can model?
+
+sealed trait MatchPathOutcome
+case object MatchPathSuccess extends MatchPathOutcome
+case object MatchPathFailed extends MatchPathOutcome //TODO: What kind of match failures do we get that we can model?
+
+sealed trait MatchResponseStageA[A]
+
+case class MatchResponseStatus(expected: InteractionResponse, actual: InteractionResponse) extends MatchResponseStageA[MatchStatusOutcome]
+case class MatchResponseHeaders(expected: InteractionResponse, actual: InteractionResponse) extends MatchResponseStageA[MatchHeadersOutcome]
+case class MatchResponseBody(expected: InteractionResponse, actual: InteractionResponse) extends MatchResponseStageA[MatchBodyOutcome]
+
+
+sealed trait MatchRequestStageA[A]
+
+case class MatchRequestMethod(expected: InteractionRequest, actual: InteractionRequest) extends MatchRequestStageA[MatchMethodOutcome]
+case class MatchRequestPath(expected: InteractionRequest, actual: InteractionRequest) extends MatchRequestStageA[MatchPathOutcome]
+case class MatchRequestHeaders(expected: InteractionRequest, actual: InteractionRequest) extends MatchRequestStageA[MatchHeadersOutcome]
+case class MatchRequestBody(expected: InteractionRequest, actual: InteractionRequest) extends MatchRequestStageA[MatchBodyOutcome]
```

Comment:
There are actually two Free Monads in play here, one for evaluating Requests and another for Responses. Both start with an ADT describing the _Actions_ or _Steps_ that can be performed as part of the process without any implementation.
As I said, each part of the ADT is an action, but you can thing of it as a function signature like this:
`case class Foo(<insert function arguments>) extends Bar[<insert function return type>]`